### ISSUE TITLE
training regime

### ISSUE THREAD

**Issue by kzkedzierska** [bowang-lab/scGPT#34]
Hi!

This is a really cool project! 

I have a couple clarifying questions about the pretrained models. I looked over the `args.json` files for the `scGPT_human` and couple of other models, however I am still unsure what arguments were chosen for setting the models up and training. Specifically:

- were GEP and GEPC (aksa MVC) the only learning objectives? I.e. was `do_mvc` set to `True` while `do_dab=False`, `use_batch_labels=False` and `domain_spec_batchnorm=False` in the `TransformerModel` class while `forawrd` method was run with `batch_labels=None`, `CLS=False`, `CCE=False`, `MVC=True` and `ECS=False`?
- was the loss set to a sum of the MSE calculated for GEP and GEPC?
- and what were the values of `explicit_zero_prob`, `mvc_decoder_style`, `fast_transformer_backend`, `pre_norm` parameters?

I will be grateful for any help you could provide in understanding the training procedure better.

**Comment by subercui** [comment]
Hi @kzkedzierska thank you for your interest!

In general, anything listed in that `args.json` within the pretraining model folder is about the setting during pre-training. I find we haven't made this clear in the readme, so just some additional explanation here. 

In the fine-tuning stage, so far, we only use the `args.json` to update the model size parameters when loading the model. During fine-tuning, you can update the configurations based on different tasks and needs. You can find the default setting in the tutorials and examples we've provided.

Regarding your question:

1. Yes, we don't use other learning objectives in the pre-training stage. BTW, those were some legacy names in the settings when we did the pretraining, so they might differ from the names we used later in the tutorials. We are wrapping up and hopefully will release the code for pre-training and data collection by this weekend. Hopefully, those settings would be more self-explaining then.

2. The loss was set to the mean of the MSE. You may find it here in the [loss.py](https://github.com/bowang-lab/scGPT/blob/main/scgpt/loss.py)

3. During pre-training, the `explicit_zero_prob = False` , `mvc_decoder_style = "inner product"`, `fast_transformer_backend = flash`, `pre_norm = False`. We usually use these default values. Most of them were options we tested during development, and then we chose to go with the default settings.

4. [0.25, 0.5, 0.75] was selected randomly during pretraining.

**Comment by kzkedzierska** [comment]
Those are very good questions!

The `"MVC": true` (I believe MVC [masked value prediction for cell embedding] == GEPC) arg in `args.json` for scGPT human would suggest to me that training is done with GEPC on. 

Based on the figure S1:

![](https://www.biorxiv.org/content/biorxiv/early/2023/07/02/2023.04.30.538439/F7.large.jpg?width=800&height=600&carousel=1)

and the following fragments from the paper:

> As illustrated in [Supplementary Figure S1A](https://www.biorxiv.org/content/10.1101/2023.04.30.538439v2.full#F7), during training, we randomly pick a proportion of the genes as unknown so their expression values are omitted in the input. Attention is only applied between the known genes and the query unknown gene itself, but not onto the positions of other unknown genes. 

> The inference steps are illustrated in [Supplementary Figure S1B](https://www.biorxiv.org/content/10.1101/2023.04.30.538439v2.full#F7). During the **inference** for cell-prompt generation, scGPT generates all genome-wide gene expression conditioned on the specific cell types. A trained cell embedding is inputted at the first position representing the cell type condition. The whole generation process of thousands of gene expression values is conducted in K iterative steps (i.e., K = 3 steps in [Supplementary Figure S1B](https://www.biorxiv.org/content/10.1101/2023.04.30.538439v2.full#F7)). (..) The gene-prompt generation works similarly in an iterative manner.

I understood that during training the prompting is done once for GEP ([here](https://github.com/bowang-lab/scGPT/blob/main/scgpt/model/model.py#L333-L351)) and once for GEPC (with `mvc_decoder` [here](https://github.com/bowang-lab/scGPT/blob/main/scgpt/model/model.py#L966-L976)). Then the loss is the sum of loss for GEP and GEPC objectives. 

It would be great to hear from @subercui on whether my thinking is correct!

**Comment by kzkedzierska** [comment]
Also, on the topic of the generative attention - is the masking of the genes provided with the `src_key_padding_mask`?

**Comment by IFLY-D** [comment]
> Hi!
> 
> This is a really cool project!
> 
> I have a couple clarifying questions about the pretrained models. I looked over the `args.json` files for the `scGPT_human` and couple of other models, however I am still unsure what arguments were chosen for setting the models up and training. Specifically:
> 
> * were GEP and GEPC (aksa MVC) the only learning objectives? I.e. was `do_mvc` set to `True` while `do_dab=False`, `use_batch_labels=False` and `domain_spec_batchnorm=False` in the `TransformerModel` class while `forawrd` method was run with `batch_labels=None`, `CLS=False`, `CCE=False`, `MVC=True` and `ECS=False`?
> * was the loss set to a sum of the MSE calculated for GEP and GEPC?
> * and what were the values of `explicit_zero_prob`, `mvc_decoder_style`, `fast_transformer_backend`, `pre_norm` parameters?
> 
> I also saw that mask_ratio is a list with [0.25, 0.5, 0.75]. Was the value selected at random for each sample or was it set to one of those?
> 
> I will be grateful for any help you could provide in understanding the training procedure better.

hello,I cannot find args.json.can you tell me?

**Comment by clairecheng0906** [comment]
@kzkedzierska Instead of using flash-attn, nn.MHA from torch can be used with key_padding_mask and attn_mask. For my understanding, the unknown genes are masked by attn_mask. The padding position is masked by key_padding_mask.
